### PR TITLE
For encoded convolution, add check for when min_width would have been larger than in_width

### PR DESCRIPTION
**Pull request by jmitrevs**
# Description

In the encoded convolution, the case when the in_width is smaller than the nominal min_width needs to be handled as a special case. This attempts to fix this situation. It has been been minimally tested in a conv1d case where in_width = 11 while the kernel width is 9. Additional tests need to be made. 

## Type of change

For a new feature or function, please create an issue first to discuss it
with us before submitting a pull request.

Note: Please delete options that are not relevant.

- [x] Bug fix (non-breaking change that fixes an issue)

## Tests

`test_conv1d_narrow.py` fails in main, but passes after this fix. Similarly with `test_conv2d_narrow.py`.

## Checklist

- [x] I have read the [guidelines for contributing](https://github.com/fastmachinelearning/hls4ml/blob/master/CONTRIBUTING.md).
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix is effective or that my feature works.

### CODE REVIEW

**Comment by jmitrevs** [comment]
This has been tested for 1D, and could be merged in the more restricted sense, but it has not been tested for 2D. I think planning should take into account the future of encoded.

**Comment by jmitrevs** [comment]
I will restrict this PR to the 1D case because I think that's the only one that would be important. If we fix it for 2D, we can make a different PR.

**Review by vloncar — CHANGES_REQUESTED** [review]

File: hls4ml/templates/vivado/nnet_utils/nnet_conv2d_stream.h [hunk 1]
```diff
@@ -14,13 +14,13 @@ void compute_scaled_indices_2d(
     const unsigned w_idx,
     ap_uint<CONFIG_T::filt_height * CONFIG_T::filt_width> *pixel_idx
 ) {
-    const unsigned sh_idx = scale_index<CONFIG_T::filt_height, CONFIG_T::stride_height, CONFIG_T::in_height>(h_idx);
+    const unsigned sh_idx = scale_index<CONFIG_T::filt_height, CONFIG_T::stride_height, CONFIG_T::in_height, CONFIG_T::min_width>(h_idx);
```
[the remark this answers is not in the document]
Comment (jmitrevs) [reply hls4ml/templates/vivado/nnet_utils/nnet_conv2d_stream.h]: Fixed.

File: test/pytest/test_conv1d_narrow.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
+from hls4ml.converters.keras_to_hls import keras_to_hls
```
[the remark this answers is not in the document]
Comment (jmitrevs) [reply test/pytest/test_conv1d_narrow.py]: Fixed.

File: test/pytest/test_conv1d_narrow.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+import numpy as np
+from sklearn.metrics import accuracy_score
+import tensorflow as tf
+from tensorflow.keras.models import model_from_json, Sequential
+from tensorflow.keras.layers import Conv1D
+import yaml
+from pathlib import Path
+
+test_root_path = Path(__file__).parent
+example_model_path = (test_root_path / '../../example-models').resolve()
```
[the remark this answers is not in the document]
Comment (jmitrevs) [reply test/pytest/test_conv1d_narrow.py]: Removed the example_model_path.

File: hls4ml/templates/vivado/nnet_utils/nnet_conv_stream.h [hunk 4]
```diff
@@ -57,10 +57,15 @@ unsigned scale_index_K_lt_S(const unsigned idx) {
     return S - K + (idx - (S - K)) % S;
 }
 
-template<unsigned K, unsigned S, unsigned W>
+template<unsigned K, unsigned S, unsigned W, unsigned min_width>
```
[the remark this answers is not in the document]
Comment (jmitrevs) [reply hls4ml/templates/vivado/nnet_utils/nnet_conv_stream.h]: I tried to modify it. Take a look.

**Comment by jmitrevs** [comment]
There are some issues still, so I am converting to a draft till I get a chance to look.

File: hls4ml/backends/fpga/fpga_backend.py [hunk 5]
```diff
@@ -414,7 +421,7 @@ def compute_conv1d_instructions(self, in_W, in_C, kernel_size=3, stride=1, pad=0
         for i in range(min_W):
             windows_int.append(int(''.join(str(p) for p in reversed(windows_bin[i])), 2))
 
-        return (min_W, windows_int)
+        return (min_W, windows_int, narrow)
```
Comment (vloncar) [inline hls4ml/backends/fpga/fpga_backend.py:424]: I don't understand why do you need to return the `narrow` state? Isn't it simpler to check `if min_w == in_W: use the narrow indexing function` and get rid of the `narrow` attribute later (the check determines the scale function to use). And could we call the narrow index function differently? You call it "narrow" but it returns unscaled index.

Comment (jmitrevs) [reply hls4ml/backends/fpga/fpga_backend.py:424]: So you are saying that if min_W == in_W you are in the "narrow" mode in all cases. I think that makes sense. Let me convince myself. 

As for the narrow name, I don't really have a strong opinion. Would "unscaled" be better?

(I'll also try to get the conflicts and precommits fixed.)

Comment (jmitrevs) [reply hls4ml/backends/fpga/fpga_backend.py]: I went ahead and made those changes. Let's see how the CI goes.

**Comment by jmitrevs** [comment]
This should now also fix the same issue with Conv2D.

**Comment by jmitrevs** [comment]
@vloncar, are you happy with the changes here as they are?

File: hls4ml/backends/fpga/fpga_backend.py [hunk 6]
```diff
@@ -461,134 +475,135 @@ def compute_conv2d_instructions(self, in_H, in_W, in_C, kernel_size=3, stride=1,
...
-                        15728640,
-                        32505856,
-                        31457280,
-                        29360128,
-                        25165824,
-                        16777216,
-                    ],
-                ),
-            )
+        if not unscaled_W and not unscaled_H:
```
Comment (vloncar) [inline hls4ml/backends/fpga/fpga_backend.py:478]: Why not? And why only in 2D?

File: hls4ml/backends/vivado/passes/conv_stream.py [hunk 7]
```diff
@@ -13,21 +15,25 @@ def transform(self, model, node):
...
                 node.get_input_variable().shape[0],
                 node.get_input_variable().shape[1],
                 node.get_attr('filt_width'),
-                node.get_attr('stride_width'))
+                node.get_attr('stride_width'),
+            )
             instructions_str = ','.join(str(i) for i in instructions)
             node.set_attr('min_width', min_w)
             node.set_attr('instructions', instructions_str)
+            node.set_attr('unscaled', node.get_input_variable().shape[0] == min_w)
```
Comment (vloncar) [inline hls4ml/backends/vivado/passes/conv_stream.py:31]: Do you really need an attribute here? Seems like a the check can also be done in the template itself. Everything can be an attribute, true, but less attributes makes it easier to keep track of them

File: hls4ml/backends/fpga/fpga_backend.py (hunk 6 above)
Comment (jmitrevs) [reply hls4ml/backends/fpga/fpga_backend.py:478]: I think you are right that the checks can be modified. I added this to fix a problem, but I can rework the checks to be more flexible.

File: hls4ml/backends/vivado/passes/conv_stream.py (hunk 7 above)
Comment (jmitrevs) [reply hls4ml/backends/vivado/passes/conv_stream.py:31]: I don't really have a strong opinion. I thought it was easier to define an attribute, to make the logic explicit, though I didn't think too much. Let me look at it once more.

Comment (jmitrevs) [reply hls4ml/backends/vivado/passes/conv_stream.py]: I agree with your proposal. I went ahead and pushed it.

File: hls4ml/backends/fpga/fpga_backend.py (hunk 6 above)
Comment (jmitrevs) [reply hls4ml/backends/fpga/fpga_backend.py]: I modified this. Take a look to see if it makes sense. The issue I had previously was that min_W and min_H sometimes weren't the hardcoded values because of the modification due to too small of an in_width or in_height.

File: hls4ml/templates/vivado/nnet_utils/nnet_conv1d_stream.h [hunk 8]
```diff
@@ -1,34 +1,30 @@
...
-void compute_scaled_indices_1d(
-    const unsigned w_idx,
-    ap_uint<CONFIG_T::filt_width> *pixel_idx
-) {
+template <class data_T, typename CONFIG_T>
+void compute_scaled_indices_1d(const unsigned w_idx, ap_uint<CONFIG_T::filt_width> *pixel_idx) {
     unsigned wp_idx = w_idx * (data_T::size / CONFIG_T::n_chan);
 
-    ComputeIndex: for (unsigned p = 0; p < data_T::size / CONFIG_T::n_chan; p++) {
+ComputeIndex:
```
Comment (vloncar) [inline hls4ml/templates/vivado/nnet_utils/nnet_conv1d_stream.h:14]: @jmduarte Any idea why the code formatter decided to use one level less indented for the loop labels? Looks a bit ugly. Can we set the indentation to be the same level as the loop?

Comment (jmitrevs) [reply hls4ml/templates/vivado/nnet_utils/nnet_conv1d_stream.h:14]: @jmduarte is the expert on the formatting. All of it was automatically done.

Comment (jmitrevs) [reply hls4ml/templates/vivado/nnet_utils/nnet_conv1d_stream.h:14]: Maybe that's the style, where the labels are one tab stop to the left? If it's consistent, it's not too bad.

**Comment by jmitrevs** [comment]
Is there anything left to do here before it's merged?

**Review by vloncar — APPROVED** [review]
